src/Host/Client/Impl/Host/LocalRHostConnector.cs

```diff
@@ -31,134 +33,150 @@ internal sealed class LocalRHostConnector : IRHostConnector {
             TimeSpan.FromSeconds(5);
 #endif
 
+        private static readonly NetworkCredential _credentials = new NetworkCredential("RTVS", Guid.NewGuid().ToString());
+
+        private readonly string _name;
         private readonly string _rhostDirectory;
         private readonly string _rHome;
-        private readonly LinesLog _log;
+        private readonly BinaryAsyncLock _connectLock = new BinaryAsyncLock();
 
-        public LocalRHostConnector(string rHome, string rhostDirectory = null) {
-            _rhostDirectory = rhostDirectory;
-            _rHome = rHome;
-            _log = new LinesLog(FileLogWriter.InTempFolder("Microsoft.R.Host.BrokerConnector"));
-        }
-
-        public async Task<RHost> Connect(string name, IRCallbacks callbacks, string rCommandLineArguments = null, int timeout = 3000, CancellationToken cancellationToken = new CancellationToken()) {
-            await TaskUtilities.SwitchToBackgroundThread();
+        private Process _brokerProcess;
+        private bool _isConnected;
 
-            var rhostDirectory = _rhostDirectory ?? Path.GetDirectoryName(typeof(RHost).Assembly.GetAssemblyPath());
-            rCommandLineArguments = rCommandLineArguments ?? string.Empty;
+        public LocalRHostConnector(string name, string rHome, string rhostDirectory = null)
+            : base(InterpreterId) {
 
-            string rhostExe = Path.Combine(rhostDirectory, RHostExe);
-            string rBinPath = Path.Combine(_rHome, RBinPathX64);
+            _name = name;
+            _rhostDirectory = rhostDirectory ?? Path.GetDirectoryName(typeof(RHost).Assembly.GetAssemblyPath());
+            _rHome = rHome;
+        }
 
-            if (!File.Exists(rhostExe)) {
-                throw new RHostBinaryMissingException();
+        public override void Dispose() {
+            if (!DisposableBag.TryMarkDisposed()) {
```

Comment:
If you have a `DisposableBag`, you don't have to override `Dispose` method. You can create create method `KillBrokerProcess` and then call `DisposableBag.Add(KillBrokerProcess)` in constructor or in `ConnectToBrokerWorker` after `_brokerProcess` field is assigned.